Replace OIDN denoiser in Lightmapper with a JNLM denoiser compute shader.

DarioSamo opened the pull request:
Original proposal: https://github.com/godotengine/godot-proposals/issues/7640

## Background

"Nonlinearly Weighted First-order Regression for Denoising Monte Carlo Renderings" (https://benedikt-bitterli.me/nfor/nfor.pdf) mentions in its opening chapter the possibility of using **Non-Local Means** with **Joint Filtering**, which has a fairly simple implementation over at https://github.com/ManuelPrandini/YoctoImageDenoiser. However, upon reviewing the implementation I found a few minor corrections that needed to be made and decided having the algorithm run in a compute shader would definitely yield far better results when it comes to runtime.

As mentioned in the proposal, the current implementation of OIDN is fairly outdated, and most importantly, **extremely slow due to being forced to use a single CPU thread**. On the scene I used for testing this out, **the overall rendering time went down from 3 minutes to just over 1 minute** in an RTX 3090 Ti with an 8K lightmap.

While it is not expected for the visual quality to be able to compete due to the nature of the algorithms (one is just a compute shader, the other uses an entire neural network and definitely understands the topic far more than I do), I think overall the result is fairly competitive and this might prove to be a decent fallback if we choose to go via the route of only using OIDN via the CLI.

One interesting advantage JNLM has over OIDN here is that it can use both the albedo and normal inputs to help reduce artifacts like seams and help improve the denoising and filtering process significantly. This could potentially be further improved if we can think of more auxiliary inputs we can use to help the algorithm out.

## Lighting Only

### Raw
![RAW_LIGHTING_ONLY](https://github.com/godotengine/godot/assets/538504/9bb28caa-5875-4979-a21c-256ed403d88b)

### JNLM
![NLM_LIGHTING_ONLY](https://github.com/godotengine/godot/assets/538504/d6719e38-39ed-47f4-a366-ba6cfee7b53f)

### OIDN
![OIDN_LIGHTING_ONLY](https://github.com/godotengine/godot/assets/538504/1db2b58b-262a-483e-88ec-b0eba44973c9)

## Composite Result

### Raw
![RAW_COMPLETE](https://github.com/godotengine/godot/assets/538504/72cc7620-530e-4f29-be8d-20c8a37e6ea6)

### JNLM
![NLM_COMPLETE](https://github.com/godotengine/godot/assets/538504/bf43f85f-5f97-482d-a507-9ae9b32271ed)

### OIDN
![OIDN_COMPLETE](https://github.com/godotengine/godot/assets/538504/5510ad0d-a7e0-472d-b74e-62c0f2b7d2c9)

## Feedback

My reason for the PR is that I think it'd be interesting if people could test the branch out and see if they achieve decent enough results for this implementation to act as the future fallback when OIDN is not installed. However, seeing as it got very close, I think it would be reasonable to also experiment with this algorithm and see if with some tweaks the result could be even closer than what I initially got.

For that I'd like to point out the following code in the shader itself:

https://github.com/DarioSamo/godot/blob/ed6a4e53e7420e93ca07bd48960e6eebe00c2f67/modules/lightmapper_rd/lm_compute.glsl#L753-L777
```glsl
	// Half the size of the patch window around each pixel that is weighted to compute the denoised pixel.
	// A value of 1 represents a 3x3 window, a value of 2 a 5x5 window, etc.
	const int HALF_PATCH_WINDOW = 4;

	// Half the size of the search window around each pixel that is denoised and weighted to compute the denoised pixel.
	const int HALF_SEARCH_WINDOW = 10;

	// For all of the following sigma values, smaller values will give less weight to pixels that have a bigger distance
	// in the feature being evaluated. Therefore, smaller values are likely to cause more noise to appear, but will also
	// cause less features to be erased in the process.

	// Controls how much the spatial distance of the pixels influences the denoising weight.
	const float SIGMA_SPATIAL = 5.0f;

	// Controls how much the lightmap color distance of the pixels influences the denoising weight.
	const float SIGMA_COLOR = 0.1f;

	// Controls how much the albedo color distance of the pixels influences the denoising weight.
	const float SIGMA_ALBEDO = 1.0f;

	// Controls how much the normal vector distance of the pixels influences the denoising weight.
	const float SIGMA_NORMAL = 0.5f;

	// Strength of the filter. The original paper recommends values around 10 to 15 times Sigma parameters.
	const float FILTER_VALUE = 15.0f * SIGMA_COLOR;
```

Some of the feature losses such as shadows and some particular areas as shown in the other picture might be fixable by playing around with these values and finding a sweet spot. It'd be preferable to not have to expose them to end-users as they can be values that are hard to understand what direct impact they have, but it could be evaluated to do so if it's deemed necessary to fit the needs of other scenes.

## TODO

- [x] Make strength configurable with a slider.
- [x] Proposal: Remove OIDN and the denoiser module.

DarioSamo committed:
Replace OIDN denoiser with a JNLM denoiser compute shader implementation.

AThousandShips commented on modules/lightmapper_rd/lm_compute.glsl:
```diff
@@ -735,4 +740,115 @@ void main() {
+
+	imageStore(dest_light, ivec3(atlas_pos, params.atlas_slice), vec4(denoised_rgb, input_color.a));
+#endif
+}
```
```suggestion
}

```
Missing line termination

Calinou commented:
Tested locally, it works as expected. The new denoiser does its job much faster than OIDN.

The scene below uses fully baked omni and spot lights (**Static** bake mode).

### Without textures

No denoiser | OIDN | JNLM *(this PR*)
-|-|-
![Screenshot_20230915_002619 webp](https://github.com/godotengine/godot/assets/180032/64c2c65c-5f83-4b70-909f-95bdd44e953c) | ![Screenshot_20230915_003518 webp](https://github.com/godotengine/godot/assets/180032/2f3a82ff-b9e6-4f46-ba47-250868b6c848) | ![Screenshot_20230915_003757 webp](https://github.com/godotengine/godot/assets/180032/cef114f3-32f2-44ca-9c4f-e053104178e5)

### With textures

No denoiser | OIDN | JNLM *(this PR*)
-|-|-
![Screenshot_20230915_002627 webp](https://github.com/godotengine/godot/assets/180032/17b58e06-e162-426e-942a-bfbda9cd7852) | ![Screenshot_20230915_003524 webp](https://github.com/godotengine/godot/assets/180032/1d9e39a2-dfcb-497c-93dd-42104067239f) | ![Screenshot_20230915_003803 webp](https://github.com/godotengine/godot/assets/180032/107ee70a-1d2c-46f9-b824-46e8d4dc6600)

It's a bit too aggressive on shadows – I wonder if the default tuning could be a bit more conservative, as textures do a good job at hiding the remaining noise anyway.

DarioSamo commented:
It would be interesting to modify these two values to see if they help out with the shadows:

- Reducing SIGMA_COLOR will give more weights to the colors closer in value. Maybe it's a bit too high so it's allowing colors in a similar range to be denoised too easily.
- FILTER_VALUE should have a pretty direct correlation to show strong the denoiser is.

I agree adjusting the strength might be easily doable by using the quality preset setting we already have for the ray count. That'd allow us to adjust the filtering to be stronger or weaker based on the expected noise.

DarioSamo commented:
Also noticed a bug that slipped by that invalid pixels where getting added to the filter. This fixes the extra darkness a bit.

Before
![NLM_LIGHTING_ONLY](https://github.com/godotengine/godot/assets/538504/7d79b145-913e-405a-8933-86de189adbb3)

After
![NLM_LIGHTING_ONLY_BUGFIX](https://github.com/godotengine/godot/assets/538504/d4d45460-8886-4c7b-bdfd-76401e714619)

wareya commented:
Looking at Calinou's screenshots makes me feel like JNLM breaks and just does a giant blur in a lot of situations where it probably shouldn't; the part under the stairs in the first screenshot set stands out a lot in a bad way.

DarioSamo commented:
> Looking at Calinou's screenshots makes me feel like JNLM breaks and just does a giant blur in a lot of situations where it probably shouldn't; the part under the stairs in the first screenshot set stands out a lot in a bad way.

I agree the feature loss is a bit much, but I think it's fixable by tweaking the one of the constants I mentioned before, so I'd be interested in how much lower the color range would need to be so the shadows are not lost but a reasonable amount of noise is reduced.

DarioSamo commented:
To provide an example, here's how much the filtering is affected by changing the filter value.

This one has a pretty direct result so maybe we could expose it as a setting next to the "Use Denoiser" toggle as well.

1.0
![NLM_LIGHTING_ONLY_FILTER_01](https://github.com/godotengine/godot/assets/538504/633bbb24-a177-4cad-a8d1-0cfd89b92690)

5.0
![NLM_LIGHTING_ONLY_FILTER_05](https://github.com/godotengine/godot/assets/538504/b7c0f665-6601-4b79-b0c2-29f289235b77)

10.0
![NLM_LIGHTING_ONLY_FILTER_10](https://github.com/godotengine/godot/assets/538504/4304f1dd-12d4-4ade-9e8f-4ac1b23d01d4)

15.0
![NLM_LIGHTING_ONLY_FILTER_15](https://github.com/godotengine/godot/assets/538504/fe228056-6d97-40d7-97c6-4a34cc5f2b39)

DarioSamo commented:
Also reducing the normal bandwidth drastically can fix the banding artifacts visible on the pillars and the carpets, it was definitely too high for its intended purpose.

![NLM_LIGHTING_ONLY_FILTER_10](https://github.com/godotengine/godot/assets/538504/9b5a8fd9-2d93-430d-8259-ee6674b97be6)
![NLM_LIGHTING_CHANGES](https://github.com/godotengine/godot/assets/538504/139cedc9-870b-468e-af52-7c3f1d206e78)

DarioSamo commented:
Exposed this new setting that controls the filter value directly so we can tweak it if necessary.

![image](https://github.com/godotengine/godot/assets/538504/96afe4ec-c2b5-40a3-a868-f2ae8b5686ac)

Calinou commented:
**Testing project:** https://0x0.st/HO8p.zip

I've tested the latest revision of this PR on the same testing project as above (`complex/complex.tscn`). On the staircase's shadows, changing the denoiser strength doesn't seem to do much:

Denoiser off | Denoiser 0.1 | Denoiser 5.0 | Denoiser 20.0
-|-|-|-
![Screenshot_20230916_023501_denoiser_off webp](https://github.com/godotengine/godot/assets/180032/42e62185-674a-4bc0-81d5-67caa4561b42) |
![Screenshot_20230916_023539_denoiser_0 1 webp](https://github.com/godotengine/godot/assets/180032/28be3c88-d14c-45f3-a990-61d8542e66e1) |
![Screenshot_20230916_023603_denoiser_5 webp](https://github.com/godotengine/godot/assets/180032/6874737e-cf9f-4204-ab5c-b14c758e823e) |
![Screenshot_20230916_023647_denoiser_20 webp](https://github.com/godotengine/godot/assets/180032/67c7882f-f01f-4bd7-b17b-810c41276591) |

Other shadows in the scene fare better at lower denoiser levels, keeping a good level of detail while not exhibiting too much visible noise when textures are added to the mix:

![image](https://github.com/godotengine/godot/assets/180032/b26e6714-8adf-4f5a-a08f-f281939f12f9)

Calinou reviewed (changes requested):
The **Denoise Strength** property in LightmapGI should be hidden when **Use Denoiser** is disabled, as it has no effect in that case.

This can be done by modifying `LightmapGI::_validate_property()` (see the existing examples there), and changing `LightmapGI::set_use_denoiser()` to call `notify_property_list_changed();` at the end of the method.

mrjustaguy commented:
A suggestion with the lightmapper in general, wouldn't it make sense to store a Non-denoised lightmap, and have the denoiser make a copy of it and save a second denoised result that gets overwritten every time denoiser settings change? it'd speed up Iteration times a ton when it comes to changing the Denoising Settings, and might make it more viable to add more settings to play around with.

akien-mga commented:
@DarioSamo If we proceed we removing OIDN as discussed, here's a commit that does it, removing all references also in license details and docs:
https://github.com/akien-mga/godot/commit/941222205690ecd6a4a6d970d2a26cdac86a5784

-115k LOC, _nice_!

DarioSamo commented:
> A suggestion with the lightmapper in general, wouldn't it make sense to store a Non-denoised lightmap, and have the denoiser make a copy of it and save a second denoised result that gets overwritten every time denoiser settings change? it'd speed up Iteration times a ton when it comes to changing the Denoising Settings, and might make it more viable to add more settings to play around with.

That'd double up the result of storing lightmaps, so it does come with the tradeoff of having to store double the textures as well as coming up with a new naming scheme for them and possibly file format changes.

I think the ideal goal here would be to avoid having to change these settings as much as possible or just have very few of them, as we don't want to impact the workflow of people significantly.

> @DarioSamo If we proceed with removing OIDN as discussed on RC, here's a commit that does it, removing all references also in license details and docs: [akien-mga@9412222](https://github.com/akien-mga/godot/commit/941222205690ecd6a4a6d970d2a26cdac86a5784)
> 
> -115k LOC, _nice_! Feel free to grab ownership and those sweet sweet negative LOC stats :D

Ah sweet, I had a similar commit of my own but I missed the COPYRIGHT.TXT file. I'll just grab yours instead.

I will proceed to make some changes and see if we can restore those shadows in Calinou's scene somehow.

mrjustaguy commented:
it could be an editor setting, and you'd just call the OG one something like `original` or `noisy` or `base` and it could be hidden away in editor cache, with the denoised being what people get in their project files.

DarioSamo commented:
Did @Calinou's change and also changed what variable of the denoiser the strength was linked to.

While it can't exactly match OIDN's result, it is significantly improved by adjusting the strength on according to the level of contrast in the scene.

![NLM_BANDWIDTH](https://github.com/godotengine/godot/assets/538504/5d2befa7-057a-4f45-a296-13a72cdc78bf)

Taking it out of drafts so we can start testing it out more in detail. I don't think the design will change significantly at this point, but we might need to evaluate whether an additional setting is needed or not to have better control over it. I'd rather keep one single slider if we can.

DarioSamo committed:
Remove denoise module and thirdparty OIDN.

This is replaced by a much lighter weight and faster JNLM denoiser. OIDN is still much more accurate, and may be provided as an optional backend in the future, but the JNLM denoiser seems good enough for most use cases and removing OIDN reduces the build system complexity, binary size, and build times very significantly.

clayjohn commented on modules/lightmapper_rd/lm_compute.glsl:
```diff
@@ -735,4 +749,118 @@ void main() {
+	// Joint Non-local means (JNLM) denoiser.
+	// Based on YoctoImageDenoiser's JNLM implementation with corrections from "Nonlinearly Weighted First-order Regression for Denoising Monte Carlo Renderings".
+	// <https://github.com/ManuelPrandini/YoctoImageDenoiser/blob/06e19489dd64e47792acffde536393802ba48607/libs/yocto_extension/yocto_extension.cpp#L207>
+	// <https://benedikt-bitterli.me/nfor/nfor.pdf>
```
We should include the full copyright notice. While its kind of awkward here, I think this chunk of code likely counts as a "substantial portion". We would also need to include the copyright info in thirdparty I beleive CC @akien-mga 
```
// LICENSE:
//
// Copyright (c) 2020 -- 2020 Fabio Pellacini
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.
```

akien-mga replied:
For the copyright info, you'd add it like this:
```diff
diff --git a/COPYRIGHT.txt b/COPYRIGHT.txt
index 500eb84d08..9abf526209 100644
--- a/COPYRIGHT.txt
+++ b/COPYRIGHT.txt
@@ -63,6 +63,11 @@ Copyright: 2011, Ole Kniemeyer, MAXON, www.maxon.net
  2007-2014, Juan Linietsky, Ariel Manzur
 License: Expat and Zlib
 
+Files: ./modules/lightmapper_rd/lm_compute.glsl
+Comment: Joint Non-Local Means (JNLM) denoiser
+Copyright: 2020, Fabio Pellacini
+License: Expat
+
 Files: ./platform/android/java/lib/aidl/com/android/*
  ./platform/android/java/lib/res/layout/status_bar_ongoing_event_progress_bar.xml
  ./platform/android/java/lib/src/com/google/android/*
```

If the code was substantially modified for Godot, then you could also add the typical Godot copyright:
```diffdiff --git a/COPYRIGHT.txt b/COPYRIGHT.txt
index 500eb84d08..7edda06c1b 100644
--- a/COPYRIGHT.txt
+++ b/COPYRIGHT.txt
@@ -63,6 +63,13 @@ Copyright: 2011, Ole Kniemeyer, MAXON, www.maxon.net
  2007-2014, Juan Linietsky, Ariel Manzur
 License: Expat and Zlib
 
+Files: ./modules/lightmapper_rd/lm_compute.glsl
+Comment: Joint Non-Local Means (JNLM) denoiser
+Copyright: 2020, Fabio Pellacini
+ 2014-present, Godot Engine contributors
+ 2007-2014, Juan Linietsky, Ariel Manzur
+License: Expat
+
 Files: ./platform/android/java/lib/aidl/com/android/*
  ./platform/android/java/lib/res/layout/status_bar_ongoing_event_progress_bar.xml
  ./platform/android/java/lib/src/com/google/android/*
```

DarioSamo replied:
It was definitely substantially modified. The license in the original repo is MIT. Do we have to specify expat or MIT?

Would we be fine with just adding this copyright notice or do we have to include the license in the code like Clay mentioned?

akien-mga replied:
The `COPYRIGHT.txt` file follows a Debian format and they use "Expat" for MIT, as they deem the MIT name too ambiguous (there's multiple versions of it with slightly different language). So it should use `Expat`.

And yeah I think it's good to include the full license in that function for good measure.

DarioSamo replied:
Done.

atirut-w commented:
Not sure if this counts as off-topic, but I'm more in favor of this than https://github.com/godotengine/godot-proposals/issues/7640 mainly for being in-house and not relying on something with weird deps.

DarioSamo commented:
> Not sure if this counts as off-topic, but I'm more in favor of this than [godotengine/godot-proposals#7640](https://github.com/godotengine/godot-proposals/issues/7640) mainly for being in-house and not relying on something with weird deps.

I think the original proposal makes sense in the context it was made, we just weren't aware JNLM could get this close of a result to probably not warrant the need of making the fallback. @Calinou agreed we can probably ignore the need of making that mechanism, but we can keep it in mind in case anyone finds this too much of a downgrade.

mrjustaguy commented:
IMO we shouldn't really bother with OIDN fallback unless there are plenty of people complaining about it, given the situation, as it'd just be adding on maintenance cost without much of a benefit considering the comparative quality and simplicity of JNLM... There are simply many other areas that make more sense to be spending dev time on vs maintaining a niche that you already have a good alternative to..

But we will see the reactions to OIDN replacement as it comes, but I doubt it'll be missed much

atirut-w commented:
Adding this since no one seems to have mentioned it: directional lightmaps seems to not be denoised. Either OP forgot to, or there's a bug.

|Normal|Directional|
|-|-|
|![image](https://github.com/godotengine/godot/assets/25323231/c2f25bad-77d0-47e8-bf3a-a9d787cd734d)|![image](https://github.com/godotengine/godot/assets/25323231/ad2a3aee-a1a2-4e32-92fe-5617e4656682)|

DarioSamo commented:
> As a side note, lightmap seams are very visible with the default denoiser strength. I tweaked it a bit, but can't get a result where there are no seams and splotches. Enabling directional seems to alleviate the problem a little bit.

Seams should be easily fixable considering the level of control we got over the algorithm, I have to imagine the UV2 in this scene is being a problem for them to appear at all. Could you share that test scene?

As for directional, I haven't checked how well it applies to those, we might need to tweak it depending on how directional lightmaps actually look in their raw form (I'm not familiar with how the denoised result looks in OIDN either with those).

atirut-w commented:
> Could you share that test scene?

[Here](https://github.com/godotengine/godot/files/12708734/lightmap_seams.zip)

DarioSamo commented:
@atirut-w I'm pretty confused about the scene you just shared. I think I see the seams but they're also visible without the denoiser. Are you sure they're even related?

DarioSamo commented:
As for spherical harmonics, I've confirmed that it's not that it's not being denoised, but rather the strength is just pretty weak:

No Denoiser
![SPHERICAL_NOISE](https://github.com/godotengine/godot/assets/538504/524717ed-8527-4fda-9c2b-7b4646fb06a9)

JNLM
![SPHERICAL_JNLM](https://github.com/godotengine/godot/assets/538504/ef7b33ba-eb3e-4264-967e-46cac22be303)

OIDN
![SPHERICAL_OIDN](https://github.com/godotengine/godot/assets/538504/b590696b-1574-4e40-84ff-eef5e4f5a2d7)

I'd say OIDN is cleaning it up better but it seems like it also glitched out and lost a lot of the intensity of the result on the sides.

I'll try to see which of the input parameters can help, but as I understand this is probably caused due to the format of SH lightmaps being very different than a regular one.

Calinou commented:
> I'll try to see which of the input parameters can help, but as I understand this is probably caused due to the format of SH lightmaps being very different than a regular one.

I suppose the denoiser strength could be automatically doubled (or even multiplied by 3 or 4) if Directional is enabled.

DarioSamo commented:
I've tried increasing it to several magnitudes higher and it's not doing anything yet, so I suspect it's something else. Looking into it.

DarioSamo commented:
Found it, wrong atlas slice was being used to query from the albedo and normal. Will push a fix.

JNLM denoising a directional lightmap

![image](https://github.com/godotengine/godot/assets/538504/8564db62-655d-4e2e-8c4b-ffd24b5d443f)

atirut-w commented:
> @atirut-w I'm pretty confused about the scene you just shared. I think I see the seams but they're also visible without the denoiser. Are you sure they're even related? Can you point to what the seams are so we're in the same page?

These?:
![image](https://github.com/godotengine/godot/assets/25323231/b495cdef-c50c-416b-b741-bcfbfc4df746)

DarioSamo commented:
> > @atirut-w I'm pretty confused about the scene you just shared. I think I see the seams but they're also visible without the denoiser. Are you sure they're even related? Can you point to what the seams are so we're in the same page?
> 
> These?: ![image](https://user-images.githubusercontent.com/25323231/270504732-b495cdef-c50c-416b-b741-bcfbfc4df746.png)

This scene has really visible seams even without the denoiser.

![image](https://github.com/godotengine/godot/assets/538504/15b18c7f-9499-4c6d-b6e9-0932df1605da)

Honestly just looks like the lightmap size is way too low to me to expect any coherence between boxes that are joined together and the denoiser is working over far too few pixels.

![image](https://github.com/godotengine/godot/assets/538504/038f65d6-6a7e-4bf2-a23c-9716f8b9bd31)

An increase in the texture size also clearly helps with the coherence.

![image](https://github.com/godotengine/godot/assets/538504/25d52abc-daca-4146-8497-038c393264f4)

But really, unless you can somehow get a better result with OIDN, which I doubt because the problem here is the coherence of the texture across the surface, you should just do a better model. You won't find this behaves much better in other engines when you give the light-mapper a hard time to match the UV2 across meshes like this.

atirut-w commented:
So, in conclusion, it's caused by bad UV2?

clayjohn reviewed (approved):
Implementation looks good to me. 

In the Production team meeting we also discussed removing OIDN and agreed that we were okay removing it for now because the quality is only slightly better. We can evaluate bringing it back as an extension if there is user demand. 

I have not tested locally, but I trust the testing done by other contributors

DarioSamo commented:
> So, in conclusion, it's caused by bad UV2?

The UV2 itself isn't the problem (although you could manually alter it to mitigate it further). The construction of the scene itself just works against how a lightmapper works in general.

Each of those cubes has its own lightmap. A denoiser will only work in the context of one texture at a time. It's not gonna be possible for any denoiser to fix that unless it starts working with multiple lightmaps at once and figuring out how things align in world space. That is a really tough problem to solve.

Think of how tough it can usually be to get a texture to tile properly across seams like this: now imagine how tough it's gonna be for a lightmapper to guarantee consistency across pixels, especially at low resolution when basically one pixel covers far more than that in the screen depending on the camera distance.

Calinou reviewed (approved):
I think this is good to go now 🙂 
Great work!

YuriSizov commented:
Thanks! This is a hefty reduction in the size of the codebase! Here's to faster build times 🎉

Anutrix commented:
**JNLM** stands for **_Joint NL-Means_**.
Took me some time find the full form.

Listwon commented:
2 years ago Blender [dropped NLM](https://code.blender.org/2021/04/cycles-x/) in favor of superior OIDN and here instead of properly implementing latest version of OIDN, Godot drops it in favor of inferior denoiser algorithm? It's not even affecting runtime binaries, just the editor, so i don't get it. NLM could be an external alternative denoiser or something you're able to select and choose from other options, but not the only core tech as it gives worse results. Quality should be more favorable for the offline denoising tool or at least switchable (to get better results in the final bake).

NLM as faster (i doubt it with multithreaded implementation of OIDN), choosable alternative sounds ok.

Open Image Denoise is used by Blender and Unity.

DarioSamo commented:
> 2 years ago Blender [dropped NLM](https://code.blender.org/2021/04/cycles-x/) in favor of superior OIDN and here instead of properly implementing latest version of OIDN, Godot drops it in favor of inferior denoiser algorithm? It's not even affecting runtime binaries, just the editor, so i don't get it. NLM could be an external alternative denoiser or something you're able to select and choose from other options, but not the only core tech as it gives worse results. Quality should be more favorable for the offline denoising tool or at least switchable (to get better results in the final bake).
> 
> NLM as faster (i doubt it with multithreaded implementation of OIDN), choosable alternative sounds ok.
> 
> Open Image Denoise is used by Blender and Unity.

It's already been explained in the original proposal, you'll never have an embedded OIDN integration that works properly due to how Godot's build system works unless you want to port their custom compiler, CUDA, HIP, etc.

The purpose of this is to have an embedded alternative that works and we can have a command line option that you can set so you can use OIDN properly. The current version that was embedded was slow to the point of being unusable (single threaded). I don't see any realistic scenario on how that was viable for production the way it was.

Listwon commented:
It's not an "alternative" if it's the only core option available. Any functionality you remove from the core/move outside will lead to less Godot users using and caring about it and in turn less people interested in using engine without industry standard functionalities (because nobody or hardly anybody maintains something outside the core, which people doesn't care about; it's a vicious cycle).

Why wouldn't you make it a choosable denoiser and try to get the funding for better implementation of OIDN or simply spread awareness that it needs more love? (i know it's not that easy, but with OIDN removed from the core it's simply impossible)

It could also encourage implementing other choosable denoisers.

DarioSamo commented:
> It's not an "alternative" if it's the only core option available. Any functionality you remove from the core/move outside will lead to less Godot users using and caring about it and in turn less people interested in using engine without industry standard functionalities (because nobody or hardly anybody maintains something outside the core, which people doesn't care about; it's a vicious cycle).
> 
> Why wouldn't you make it a choosable denoiser and try to get the funding for better implementation of OIDN or simply spread awareness that it needs more love? (i know it's not that easy, but with OIDN removed from the core it's simply impossible)
> 
> It could also encourage implementing other choosable denoisers.

The plan is already laid out in the original proposal on how to go about it. It's not out of the cards to add back OIDN and it'd definitely be as part of the core as this is, it's just the way OIDN's build system works that makes it prohibitive for Godot to embed it by itself (TBB, CUDA, HIP and ISPC). It could be as simple as automatically downloading the executable from Intel when you choose the option in the lightmapper. People wouldn't have to go hunt down an unmaintained extension. I'm very much in favor of that being implemented.

It was also missing from the blog post perhaps, but it's worth mentioning this denoiser can actually fix stuff OIDN was just getting wrong or straight up breaking, as it lacks the information from normal maps and albedo which helps guide the result a lot better than your average NLM implementacion (hence the J part of the name).

bruvzg commented:
>  The effort to get OIDN working well in Godot would be monumental.

Is there any reason not to use `oidnDenoise` executable as an external tool (as an option), like we do with blender? It's using extremely simple PFM image format (basically 3 lines of text metadata and raw float image data, supporting reading/writing it should be trivial) and seems to support all features OIDN module were using.

DarioSamo commented:
> > The effort to get OIDN working well in Godot would be monumental.
> 
> Is there any reason not to use `oidnDenoise` executable as an external tool (as an option), like we do with blender? It's using extremely simple PFM image format (basically 3 lines of text metadata and raw float image data, supporting reading/writing it should be trivial) and seems to support all features OIDN module were using.

That's pretty much what [the proposal](https://github.com/godotengine/godot-proposals/issues/7640) lays out we should do, and I agree. @Calinou pointed out there might be some other things to solve when it comes to the image format at least.

Admittedly maybe this PR went out earlier than having that option back in yet, but it wouldn't take as much effort and it'd yield a far better result than choosing to make it part of Godot itself.

clayjohn commented:
> > The effort to get OIDN working well in Godot would be monumental.
> 
> Is there any reason not to use `oidnDenoise` executable as an external tool (as an option), like we do with blender? It's using extremely simple PFM image format (basically 3 lines of text metadata and raw float image data, supporting reading/writing it should be trivial) and seems to support all features OIDN module were using.

That's the current plan!

Listwon commented:
> Is there any reason not to use oidnDenoise executable as an external tool (as an option), like we do with blender?
I think the only reason could be that it's a sample implementation and the commands can change between binary releases. But it's good enough approach, allows passing extra data to OIDN (albedo, normals) and running it on multiple CPU threads or GPU.

bruvzg commented:
> I think the only reason could be that it's a sample implementation and the commands can change between binary releases.

I guess loading the dynamic lib should not have these issues (at least across the same major version), and can be done in the same way as executable (the only downside, there's no ARM64 macOS prebuild).

Here's a quickly made test for dynamic library loading - https://github.com/bruvzg/godot/tree/oidn_external_lib (seems to work, at least on macOS), library location can be set in `Editor Settings` → `Filesystem` → `Tools` → `Oidn` → `Oidn library path`.

DarioSamo commented:
> > I think the only reason could be that it's a sample implementation and the commands can change between binary releases.
> 
> I guess loading the dynamic lib should not have these issues (at least across the same major version), and can be done in the same way as executable (the only downside, there's no ARM64 macOS prebuild).
> 
> Here's a quickly made test for dynamic library loading - https://github.com/bruvzg/godot/tree/oidn_external_lib (seems to work, at least on macOS), library location can be set in `Editor Settings` → `Filesystem` → `Tools` → `Oidn` → `Oidn library path`.

That's pretty sweet, I think you'll have an easy chance for this to get accepted now that the JNLM denoiser is in.

bruvzg commented:
Not sure which option is better, so made both:

https://github.com/godotengine/godot/pull/82831 (dynamic library)
https://github.com/godotengine/godot/pull/82832 (executable)
